auth/oidc.go

```diff
@@ -439,3 +514,182 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(duration):
+				duration, err = op.runDiscoverySync()
+				if err != nil {
+					base.Errorf("openID Connect provider discovery sync ends up in error: %v", err)
+					return
+				}
+			case <-op.stopDiscoverySync:
+				base.Debugf(base.KeyAll, "Terminating OpenID Connect provider discovery sync")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// sync polls the provider metadata from discovery endpoint and refresh the ID token verifier.
+func (op *OIDCProvider) runDiscoverySync() (time.Duration, error) {
+	discoveryURL := strings.TrimSuffix(op.Issuer, "/") + OIDCDiscoveryConfigPath
+	base.Debugf(base.KeyAuth, "Syncing provider metadata from: %s", base.UD(discoveryURL))
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	var expiresAt time.Time
+	var ttl time.Duration
+	var ok bool
+	ttl, ok, err = cacheable(resp.Header)
+	if err != nil {
+		return 0, err
+	} else if ok && ttl > 0 {
+		expiresAt = time.Now().UTC().Add(ttl)
+	}
+
+	metadataChecksum, err := sha1Checksum(resp)
+	if op.metadataChecksum == metadataChecksum {
+		op.metadataChecksum = metadataChecksum
+		base.Debugf(base.KeyAuth, "No change in discovery config detected at this time, next sync will be at %v", expiresAt)
+		return op.nextSyncAfter(expiresAt), nil
+	}
+	op.metadataChecksum = metadataChecksum
+
+	var metadata ProviderMetadata
+	if err := base.JSONDecoder(resp.Body).Decode(&metadata); err != nil {
+		base.Debugf(base.KeyAuth, "Error parsing body during discovery sync: %v", err)
+		return 0, err
+	}
+
+	if metadata.Issuer != op.Issuer {
+		return 0, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, metadata.Issuer)
+	}
+	verifier := op.GenerateVerifier(&metadata, context.Background())
+	op.client.SetVerifier(verifier)
+	op.client.SetEndpoint(metadata.Endpoint())
+	return op.nextSyncAfter(expiresAt), nil
+}
+
+// nextSyncAfter returns the elapsed time between now and the next provider
+// metadata sync time. If the metadata expiry is zero or half of the duration
+// between provider metadata expiry and now is grater than 24 hours, the next
+// sync will occur in 24 hours. If half of the duration between provider metadata
+// expiry and now is grater less than 1 minute, the next sync will occur in next
+// minute. If half of the duration between provider metadata expiry and now falls
+// in between 1 minute and 24 hours, the next sync will occur in that duration.
```

Comment:
If 'expires at' is a fixed timestamp, this looks like it's going to result in increasing sync frequency as we approach that timestamp - is that actually the intended behaviour?  

e.g. if the expiry is initially 1 day in the future, are we wanting sync to run in 12h, 6h, 3h, 1.5h, etc as we approach that time?  Wouldn't it be sufficient to schedule it to run on the normal interval until expiry is closer than 24 hours, then schedule it to run 1m after expiry?